src/main/java/org/springframework/data/redis/connection/jedis/JedisStreamCommands.java

```diff
@@ -0,0 +1,496 @@
+			throw new UnsupportedOperationException("'XINFO' cannot be called in pipeline / transaction mode.");
+		}, key).get(streamInfo -> StreamInfo.XInfoStream.fromList(mapToList(streamInfo.getStreamInfo())));
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.redis.connection.RedisStreamCommands#xInfoGroups(byte[])
+	 */
+	@Override
+	public StreamInfo.XInfoGroups xInfoGroups(byte[] key) {
+		Assert.notNull(key, "Key must not be null!");
+
+		return connection.invoke().from(BinaryJedis::xinfoGroup, (k, r) -> {
+			throw new UnsupportedOperationException("'XINFO GROUPS' cannot be called in pipeline / transaction mode.");
+		}, key).get(streamGroupInfos -> {
+			List<Object> sources = new ArrayList<>();
+			streamGroupInfos.forEach(streamGroupInfo -> sources.add(mapToList(streamGroupInfo.getGroupInfo())));
+			return StreamInfo.XInfoGroups.fromList(sources);
+		});
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.redis.connection.RedisStreamCommands#xInfoConsumers(byte[], java.lang.String)
+	 */
+	@Override
+	public StreamInfo.XInfoConsumers xInfoConsumers(byte[] key, String groupName) {
+		Assert.notNull(key, "Key must not be null!");
+		Assert.hasText(groupName, "Group name must not be null or empty!");
+
+		return connection.invoke().from(BinaryJedis::xinfoConsumers, (k, v, r) -> {
+			throw new UnsupportedOperationException("'XINFO CONSUMERS' cannot be called in pipeline / transaction mode.");
+		}, key, JedisConverters.toBytes(groupName)).get(streamConsumersInfos -> {
+			List<Object> sources = new ArrayList<>();
+			streamConsumersInfos
+					.forEach(streamConsumersInfo -> sources.add(mapToList(streamConsumersInfo.getConsumerInfo())));
+			return StreamInfo.XInfoConsumers.fromList(groupName, sources);
+		});
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.redis.connection.RedisStreamCommands#xLen(byte[])
+	 */
+	@Override
+	public Long xLen(byte[] key) {
+		Assert.notNull(key, "Key must not be null!");
+
+		return connection.invoke().just(BinaryJedis::xlen, MultiKeyPipelineBase::xlen, key);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.redis.connection.RedisStreamCommands#xPending(byte[], java.lang.String)
+	 */
+	@Override
+	public PendingMessagesSummary xPending(byte[] key, String groupName) {
+		throw new UnsupportedOperationException("Jedis does not support returning PendingMessagesSummary.");
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.redis.connection.RedisStreamCommands#xPending(byte[], java.lang.String, org.springframework.data.redis.connection.RedisStreamCommands.XPendingOptions)
+	 */
+	@Override
+	public PendingMessages xPending(byte[] key, String groupName, XPendingOptions options) {
+		Assert.notNull(key, "Key must not be null!");
+		Assert.notNull(groupName, "GroupName must not be null!");
+
+		Range<String> range = (Range<String>) options.getRange();
+		byte[] group = LettuceConverters.toBytes(groupName);
+		try {
+			if (isPipelined()) {
```

Comment:
Let's use `JedisInvoker` here.

Reply:
Do you have any good suggestions? I'm doing this because the return types of the BinaryJedis::xpending MultiKeyPipelineBase::xpending two methods are not the same.

Reply:
Yikes. Instead, you could unfold the method reference into a lambda and adopt the parameters there. Also, it would make sense to file a ticket so Jedis folks can fix the method signature.

Reply:
Thanks. fixed at lastest commit. will file  a ticket on Jedis later.